docs/csharp/tutorials/attributes.md

```diff
@@ -0,0 +1,164 @@
+
+The string is being passed as an argument to an `ObsoleteAttributes` constructor, just as if you were writing `var attr = new ObsoleteAttribute("some string")`.
+
+Parameters to an attribute constructor are limited to simple types/literals: `bool, int, double, string, Type, enums, etc` and arrays of those types.
+You can not use an expression or a variable. You are free to use positional or named parameters.
+_There is a 'gotcha' here to watch out for: the C# compiler won't stop you from creating those parameters, it will only stop you from using them_.
+
+## How to create your own attribute
+
+Creating an attribute is as simple as inheriting from the `Attribute` base class.
+
+[!code-csharp[Create your own attribute](../../../samples/snippets/csharp/tutorials/attributes/Program.cs#CreateAttributeExample1)]
+
+With the above, I can now use `[MySpecial]` (or `[MySpecialAttribute]`) as an attribute elsewhere in the code base.
+
+[!code-csharp[Using your own attribute](../../../samples/snippets/csharp/tutorials/attributes/Program.cs#CreateAttributeExample2)]
+
+Attributes in the .NET base class library like `ObsoleteAttribute` trigger certain behaviors within the compiler. However, any attribute you create acts
+only as metadata, and doesn't result in any code within the attribute class being executed. It's up to you to act
+on that metadata elsewhere in your code (more on that later in the tutorial).
+
+## How to restrict attribute usage
+
+Attributes can be used on a number of "targets". The above examples show them on classes, but they can also be used on:
+
+* Assembly
+* Class
+* Constructor
+* Delegate
+* Enum
+* Event
+* Field
+* GenericParameter
+* Interface
+* Method
+* Module
+* Parameter
+* Property
+* ReturnValue
+* Struct
+
+When you create an attribute class, by default, C# will allow you to use that attribute on any of the possible attribute targets. If you want to restrict your attribute
+to certain targets, you can do so by using the `AttributeUsageAttribute` on your attribute class. That's right, an attribute on an attribute!
+
+[!code-csharp[Using your own attribute](../../../samples/snippets/csharp/tutorials/attributes/Program.cs#AttributeUsageExample1)]
+
+If you attempt to put that attribute on something that's not a class or a struct, you will get a compiler error
+like `Attribute 'MyAttributeForClassAndStructOnly' is not valid on this declaration type. It is only valid on 'class, struct' declarations`
+
+[!code-csharp[Using your own attribute](../../../samples/snippets/csharp/tutorials/attributes/Program.cs#AttributeUsageExample2)]
+
+## How to use attributes attached to a code element
+
+Attributes act as metadata. Without some outward force, they won't actually do anything.
+
+To find and act on attributes, Reflection is generally needed. I won't cover Reflection in-depth in this tutorial, but the basic
+idea is that Reflection allows you to write code in C# that examines other code.
```

Comment:
We should have a link here. I'd consider here: /docs/csharp/programming-guide/concepts/reflection